### PR TITLE
Add support to map BIOS attributes to Security Attributes

### PR DESCRIPTION
**Pull request by superm1**
This will allow clients to be able to suggest to users to change BIOS settings to try to fix security problems.

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
Depending upon how this is used it may also make sense to try to include information about what value is the correct "positive" one to use in the security attr.

For some cases it's super obvious (enable/disable) but others you gotta know (sleep state of Linux or Windows 10).

File: libfwupdplugin/fu-self-test.c [hunk 1]
```diff
@@ -2948,7 +2948,7 @@ fu_bios_attrs_load_func(void)
 	g_assert_false(ret);
 
 	/* check an attribute reads from kernel as expected by fwupd today */
-	attr = fu_context_get_bios_attr(ctx, "AMDMemoryGuard");
+	attr = fu_context_get_bios_attr(ctx, "com.thinklmi.AMDMemoryGuard");
```
Comment (hughsie) [inline libfwupdplugin/fu-self-test.c:2951]: I think this is a really good idea.

File: libfwupd/fwupd-bios-attr.c [hunk 2]
```diff
@@ -477,6 +518,12 @@ fwupd_bios_attr_to_variant(FwupdBiosAttr *self)
 			      "{sv}",
 			      FWUPD_RESULT_KEY_BIOS_ATTR_TYPE,
 			      g_variant_new_uint64(priv->kind));
+	if (priv->attr_id != NULL) {
+		g_variant_builder_add(&builder,
+				      "{sv}",
+				      FWUPD_RESULT_KEY_BIOS_ATTR_ID,
```
Comment (hughsie) [inline libfwupd/fwupd-bios-attr.c:524]: I think FWUPD_RESULT_KEY_APPSTREAM_ID works here, no?

File: libfwupd/fwupd-security-attr.c [hunk 3]
```diff
@@ -223,6 +224,46 @@ fwupd_security_attr_flag_to_suffix(FwupdSecurityAttrFlags flag)
...
+ *
+ * Gets the #FwupdBiosAttr that can be used to improve this
+ * #FwupdSecurityAttr.
+ *
+ * Returns: The unique ID used for #FwupdBiosAttr or NULL
+ *
+ * Since: 1.8.4
+ **/
+const gchar *
+fwupd_security_attr_get_bios_attr_id(FwupdSecurityAttr *self)
```
Comment (hughsie) [inline libfwupd/fwupd-security-attr.c:239]: I think fwupd_security_attr_get_bios_id is better tbh.

File: plugins/acpi-facp/fu-plugin-acpi-facp.c [hunk 4]
```diff
@@ -41,6 +41,12 @@ fu_plugin_acpi_facp_add_security_attrs(FuPlugin *plugin, FuSecurityAttrs *attrs)
 		return;
 	}
 	if (!fu_acpi_facp_get_s2i(facp)) {
+		FwupdBiosAttr *bios_attr =
+		    fu_context_get_bios_attr(fu_plugin_get_context(plugin), "SleepState");
+		if (bios_attr != NULL) {
+			fwupd_security_attr_set_bios_attr_id(attr,
+							     fwupd_bios_attr_get_id(bios_attr));
+		}
```
Comment (hughsie) [inline plugins/acpi-facp/fu-plugin-acpi-facp.c:49]: I think we do need to say what bios attr value we need to fix this. e.g. have something like `fwupd_security_attr_set_bios_value` making it clear that this value is the *desired* value, not the actual value.

File: plugins/acpi-facp/fu-plugin-acpi-facp.c [hunk 5]
```diff
@@ -41,6 +41,12 @@ fu_plugin_acpi_facp_add_security_attrs(FuPlugin *plugin, FuSecurityAttrs *attrs)
 		return;
 	}
 	if (!fu_acpi_facp_get_s2i(facp)) {
+		FwupdBiosAttr *bios_attr =
+		    fu_context_get_bios_attr(fu_plugin_get_context(plugin), "SleepState");
```
Comment (hughsie) [inline plugins/acpi-facp/fu-plugin-acpi-facp.c:45]: I think we want to get this by AppStream ID, no? e.g. the desired value from `com.lenovo.SleepState` might be `idle` and the desired value for `com.dell.SleepState` might be `s2idle`.

File: libfwupd/fwupd-bios-attr.c (hunk 2 above)
Comment (superm1) [reply libfwupd/fwupd-bios-attr.c:524]: That was my original thought. But if you want to share with security attr you need to have two different key names.

File: plugins/acpi-facp/fu-plugin-acpi-facp.c (hunk 5 above)
Comment (superm1) [reply plugins/acpi-facp/fu-plugin-acpi-facp.c:45]: Yes. This is only for Lenovo. Will fix to add the ID.

File: libfwupd/fwupd-bios-attr.c (hunk 2 above)
Comment (hughsie) [reply libfwupd/fwupd-bios-attr.c:524]: Bleh, indeed. Sorry.

File: plugins/acpi-facp/fu-plugin-acpi-facp.c (hunk 4 above)
Comment (superm1) [reply plugins/acpi-facp/fu-plugin-acpi-facp.c:49]: Right. I think we need two keys then. Like a desired value key and which BIOS setting to change key.

Comment (hughsie) [reply plugins/acpi-facp/fu-plugin-acpi-facp.c:49]: Agree.

File: libfwupd/fwupd-bios-attr.c [hunk 6]
```diff
@@ -22,10 +22,12 @@ fwupd_bios_attr_finalize(GObject *object);
 
 typedef struct {
 	FwupdBiosAttrKind kind;
+	gchar *attr_id;
```
Comment (hughsie) [inline libfwupd/fwupd-bios-attr.c:25]: hmm, i still think this one wants to be appstream_id.

Comment (hughsie) [reply libfwupd/fwupd-bios-attr.c:25]: if not, then just `id`

File: libfwupd/fwupd-bios-attr.c [hunk 7]
```diff
@@ -546,6 +645,10 @@ fwupd_bios_attr_from_key_value(FwupdBiosAttr *self, const gchar *key, GVariant *
 		fwupd_bios_attr_set_path(self, g_variant_get_string(value, NULL));
 		return;
 	}
+	if (g_strcmp0(key, FWUPD_RESULT_KEY_BIOS_ATTR_POSITIVE_VALUE) == 0) {
+		fwupd_bios_attr_set_positive_value(self, g_variant_get_string(value, NULL));
```
Comment (hughsie) [inline libfwupd/fwupd-bios-attr.c:649]: not sure i like the `positive` noun. Got any other ideas?

File: libfwupd/fwupd-bios-attr.c (hunk 6 above)
Comment (superm1) [reply libfwupd/fwupd-bios-attr.c:25]: It shouldn't be `appstream_id`, I think `id` is fine.

File: libfwupd/fwupd-bios-attr.c (hunk 7 above)
Comment (superm1) [reply libfwupd/fwupd-bios-attr.c:649]: OK so if the heuristic to set it is dropped from `fu-bios-attrs`, how about `preferred`?  Then we can set preferred only with security attributes.

Comment (hughsie) [reply libfwupd/fwupd-bios-attr.c:649]: `preferred` makes sense to me. That or `ideal` or `endorsed` or `recommended`

File: libfwupd/fwupd-bios-attr.c [hunk 8]
```diff
@@ -22,10 +22,12 @@ fwupd_bios_attr_finalize(GObject *object);
 
 typedef struct {
 	FwupdBiosAttrKind kind;
+	gchar *id;
 	gchar *name;
 	gchar *description;
 	gchar *path;
 	gchar *current_value;
+	gchar *positive_value;
```
Comment (hughsie) [inline libfwupd/fwupd-bios-attr.c:30]: preferred

File: libfwupd/fwupd-bios-attr.c [hunk 9]
```diff
@@ -495,6 +584,12 @@ fwupd_bios_attr_to_variant(FwupdBiosAttr *self)
 				      FWUPD_RESULT_KEY_DESCRIPTION,
 				      g_variant_new_string(priv->description));
 	}
+	if (priv->positive_value != NULL) {
+		g_variant_builder_add(&builder,
+				      "{sv}",
+				      FWUPD_RESULT_KEY_BIOS_ATTR_POSITIVE_VALUE,
```
Comment (hughsie) [inline libfwupd/fwupd-bios-attr.c:590]: same

File: libfwupd/fwupd-enums-private.h [hunk 10]
```diff
@@ -508,11 +508,26 @@ G_BEGIN_DECLS
...
+ **/
+#define FWUPD_RESULT_KEY_BIOS_ATTR_ID "BiosAttrId"
+/**
+ * FWUPD_RESULT_KEY_BIOS_ATTR_POSITIVE_VALUE:
+ *
+ * Result key to represent the value that would enable this attribute.
+ *
+ * The D-Bus type signature string is 's' i.e. a string.
+ **/
+#define FWUPD_RESULT_KEY_BIOS_ATTR_POSITIVE_VALUE "BiosAttrPositiveValue"
```
Comment (hughsie) [inline libfwupd/fwupd-enums-private.h:526]: same

File: libfwupd/fwupd-bios-attr.c (hunk 8 above)
Comment (superm1) [reply libfwupd/fwupd-bios-attr.c:30]: gah, forgot a few thanks.

**Review by hughsie — APPROVED** [review]
Looks good. Do you have any ideas on how this could work on the CLI? e.g. do we want a --interactive set of prompts for `fwupdmgr security`? A different PR tho.

**Comment by superm1** [comment]
> Looks good. Do you have any ideas on how this could work on the CLI? e.g. do we want a --interactive set of prompts for `fwupdmgr security`? A different PR tho.

I hadn't really given it thought until you asked.  I think at the end of showing security information there could be a prompt along the lines of:
````
fwupd could change some BIOS settings for you for the following items:
Foo
* bar->baz 
SecureBoot
*  Disabled -> Enabled

WARNING: Although changing BIOS settings may help improve your security score, it may cause problems with the operation of your system, please be prepared to enter your firmware setup to change settings back if any of these cause a problem for you.
```

**Comment by hughsie** [comment]
Ya, something exactly like that.

**Comment by hughsie** [comment]
Maybe allowlisting them for now too.

**Comment by superm1** [comment]
> Maybe allowlisting them for now too.

What do you mean?  Only certain settings are tagged to a security attribute.

**Comment by hughsie** [comment]
> What do you mean? Only certain settings are tagged to a security attribute.

Ohh, i mean we probably want bespoke translations and explanations for the super common ones, like turning on SB.

**Comment by superm1** [comment]
> > What do you mean? Only certain settings are tagged to a security attribute.
> 
> Ohh, i mean we probably want bespoke translations and explanations for the super common ones, like turning on SB.

well that's why I want to have more values filled in for the Dell stuff and some hardcoded stuff for Lenovo.  I think if we can dump all of the known lenovo attributes into a google doc and then ask the internet to fill it in we should get all the strings we need (and then some).

**Pull request merged**